BGL/include/CGAL/boost/graph/helpers.h

```diff
@@ -355,31 +356,293 @@ bool is_valid_face_descriptor( typename boost::graph_traits<FaceGraph>::face_des
+      break;
+  }
+  ++n;
+}
+valid = valid && (n == num_h);
+if ( n != num_h)
+  verr << "counting halfedges failed." << std::endl;
+
+verr << "structure is "
+   << ( valid ? "valid." : "NOT VALID.") << std::endl;
+return valid;
+}
+
+/*!
+  \ingroup PkgBGLHelperFct
+ * \brief checks the integrity of `g`.
+ * 
+ * `g` is valid if it is a valid `HalfedgeListGraph`, if it follows the rules
+ * of the `FaceListGraph` concept, and all of its associations are reciprocal.
+ * For example, `face(halfedge(f,g),g)` must be `f`.
+ * calls `is_valid_halfedge_graph()`
+ * \param g the `Graph` to test.
+ * \param verb : if `true`, the details of the check will be written in the standard output.
+ *
+ * \tparam `Graph` a model of `FaceListGraph`
+ * \return `true` if `g` is valid, `false` otherwise.
+ * 
+ * \see `is_valid_halfedge_graph()` 
+ */
+template<typename Graph>
+bool is_valid_face_graph(const Graph& g, bool verb = false)
+{
+  typedef typename boost::graph_traits<Graph>::halfedge_descriptor   halfedge_descriptor;
+  typedef typename boost::graph_traits<Graph>::face_descriptor       face_descriptor;
+  typedef typename boost::graph_traits<Graph>::faces_size_type       faces_size_type;
+  typedef typename boost::graph_traits<Graph>::halfedges_size_type   halfedges_size_type;
+  std::size_t num_f(std::distance(boost::begin(faces(g)), boost::end(faces(g)))),
+      num_h(std::distance(boost::begin(halfedges(g)), boost::end(halfedges(g))));
+  
+  //is valid halfedge_graph ?
+  bool valid=is_valid_halfedge_graph(g, verb);
+  if ( ! valid) {
+    return false;
+  }
+  Verbose_ostream verr(verb);
+    
+  // All faces.
+  faces_size_type f = 0;
+  std::size_t n = 0;
+  halfedges_size_type nb = 0;
+  BOOST_FOREACH(face_descriptor fbegin, faces(g)){
+    if(!valid)
+      break;
+    verr << "face " << f << std::endl;
+    // Pointer integrity.
+    if ( halfedge(fbegin, g) != boost::graph_traits<Graph>::null_halfedge())
+      valid = valid && (
+            face(halfedge(fbegin, g), g) == fbegin);
+    else
+      valid = false;
+    if ( ! valid) {
+      verr << "    halfedge pointer in face corrupted." << std::endl;
+      break;
     }
+    // cycle-around-face test.
+    halfedge_descriptor h = halfedge( fbegin, g);
+    if (h != boost::graph_traits<Graph>::null_halfedge()) {
+      halfedge_descriptor ge = h;
+      do {
+        verr << "    halfedge " << n << std::endl;
+        ++n;
+        h = next(h, g);
+        valid = valid && ( n <= num_h && n!=0);
+        if ( ! valid)
+          verr << "    too many halfedges around faces."
+               << std::endl;
+      } while ( valid && (h != ge));
+    }
+    ++f;
   }
-  return true;
+  if ( valid && f != num_f)
+    verr << "counting faces failed." << std::endl;
+  
+  BOOST_FOREACH(halfedge_descriptor i, halfedges(g)){
+    //counting borders
+    if ( is_border(i, g))
+      ++nb;
+    // face integrity.
+    
+    valid = valid && ( face(i, g) == face(next(i, g), g));
```

Comment:
@sloriot That should check that all halfedges in a loop point to the same face, shouldn't it ?

Reply:
Great, thanks!